lib/modules/dosomething/dosomething_global/dosomething_global.module

```diff
@@ -256,3 +273,143 @@ function dosomething_global_get_language($account, stdClass $node = NULL) {
+ * - and I'm not a privileged user (admin, editor, etc.),
+ * - and I request a bare URL (e.g., /campaigns/sample),
+ * - and there is a translation available (e.g., /voluntario/ejemplo),
+ * - then issue a 302 redirect to the translation's URL.
+ */
+function dosomething_global_redirect_node_auth() {
+  global $user;
+
+  // Don't do redirects for privileged users.
+  if (user_access('access administration menu')) {
+    return;
+  }
+
+  // User language.
+  $user_lang_code = $user->language ?: 'en';
+
+  // All languages, as an array of objects.
+  $languages = language_list();
+
+  $global_lang_code = 'en-global';
+
+  if (empty($languages[$global_lang_code])) {
+    watchdog('dosomething_global', "Can't load %lang language",
+      array('%lang' => $global_lang_code), WATCHDOG_ERROR);
+    return;
+  }
+  $language_global = $languages[$global_lang_code];
+
+  // Don't redirect for Global English speakers.
+  if ($user_lang_code == $language_global->language) {
+    return;
+  }
+
+  $node = node_load(arg(1));
+
+  if ($languages[$user_lang_code]) {
+    dosomething_global_redirect_node_language($node, $user_lang_code, 302);
+  }
+}
+
+/**
+ * Handle redirects to translated content for anonymous users. In this
+ * case:
+ *
+ * - If I request a bare URL (e.g., /campaigns/sample),
+ * - and there is a translation available (e.g., /voluntario/ejemplo),
+ * - then issue a 301 redirect to the translation's URL.
+ */
+function dosomething_global_redirect_node_anon() {
+
+  // All languages, as an array of objects.
+  $languages = language_list();
+
+  // Map GeoIP country to a system language. Default to 'en-global'.
+  $user_country_code = dosomething_settings_get_geo_country_code();
+  $target_lang_code = 'en-global';
+
+  $lang_map = variable_get('dosomething_global_language_map', '');
+  foreach ($lang_map as $lang_key => $lang) {
+    if ($lang['country'] == $user_country_code) {
+      $target_lang_code = $lang_key;
+      break;
+    }
+  }
+
+  // If we're actually supposed to show the en-global version:
+  if ('en-global' == $target_lang_code) {
+    return;
+  }
+
+  if (empty($languages[$target_lang_code])) {
+    watchdog('dosomething_global', "Can't load %lang language",
+      array('%lang' => $target_lang_code), WATCHDOG_ERROR);
+    return;
+  }
+
+  $node = node_load(arg(1));
+
+  dosomething_global_redirect_node_language($node, $target_lang_code, 301);
+}
+
+/**
+ * Given a node object, a language code (e.g., "en", "en-global", "es-mx"),
+ * and a redirect status, redirect the request to the desired translation's
+ * URL.
+ *
+ * @param $node
+ * @param $language_code
+ * @param int $redirect_status
+ */
+function dosomething_global_redirect_node_language($node, $language_code, $redirect_status = 302) {
+
+  $languages = language_list();
+
+  // Can only redirect to an published, translated version.
+  if (!empty($node->translations->data[$language_code]) && (bool)$node->translations->data[$language_code]['status']) {
```

Comment:
If the translation doesn't exist, the code after `!empty($node->translations->data[$language_code])` won't execute. If `$node->translations->data[$language_code]` isn't empty, then `['status']` will be either `1` or `0`.